In [1]:
# Advent of Code, Day 8 - Jim Carson. 
def parse(puzzle_input):
    r = []
    with open(puzzle_input,"r") as fp:
        f = fp.read()
    lines = f.splitlines()
    for l in lines:
        s, d = l.split("|")
        signals = s.split()
        displays = d.split()
        r.append((signals, displays))
    return r

Each digit of a seven-segment display is rendered by turning on or off any of seven segments named a through g:

![7-segment Display](7_segment.png)

Each entry consists of ten unique signal patterns, a | delimiter, and finally the four digit output value. Within an entry, the same wire/segment connections are used (but you don't know what the connections actually are). The unique signal patterns correspond to the ten different ways the submarine tries to render a digit using the current wire/segment connections. 

acedgfb cdfbe gcdfa fbcad dab cefabd cdfgeb eafb cagedb ab **|**
*cdfeb fcadb cdfeb cdbaf*


In [2]:
data = parse("input_files/day08.txt")
num_1478 = 0
for i in data:
    for j in i[1]:
        if len(j) in [2,3,4,7]:
            num_1478 += 1
print(num_1478)

470


In the output values, how many times do digits 1, 4, 7, or 8 appear?
**470**

In [3]:
#
# This works by logically deducing segments based on length and existing segments in other, known
# numbers.  Because of timing, there's no error checking.  Rather, you have to call it by length,
# skipping length(5) numbers until length(6) are identified.  (I couldn't figure out how to sort a 
# list by length 2, 3, 4, 7, 6, 5.)
#
def identify_digit(signal, known):
    L = len(signal)
    S = set(list(signal))
    if L == 2:
        return(1,S)
    if L == 3:
        return(7,S)
    if L == 4:
        return(4,S)
    if L == 7:
        return(8,S)
    # Ideally, there would be a check here that 1, 4, and 7 are already identified.
    if L == 6:
        if known[7].issubset(S):
            if known[4].issubset(S):
                return(9,S)
            return(0,S)
        return(6,S)
    # Ideally, there would be a check here that 9 and 7 are already identified.
    if L == 5:
        if S.issubset(known[9]) and not known[7].issubset(S):
            return(5,S)
        elif known[7].issubset(S):
            return(3,S)
        return(2,S)
    # This should never happen as the problem is stated
    return(None)

def get_key(val, d):
    for k, v in d.items():
        if val == v:
            return k

# Sort the list by length, but in this order: 2, 3, 4, 7, 6, 5.
def prioritize(signal):
    for i in sorted(signal, key=len):
        if len(i) != 5:
            yield i
    for i in sorted(signal, key=len):
        if len(i) == 5:
            yield i

def deduce_numbers(signal, display):
    known = {}
    for i in prioritize(signal):
        num, S = identify_digit(i, known)
        known[num] = S
    num = []
    for i in display:
        num.append(get_key(set(list(i)),known))
    return int(''.join(map(str, num)))

For each entry, determine all of the wire/segment connections and decode the four-digit output values. What do you get if you add up all of the output values?  **989396**

In [4]:
outputs = []
for signal, display in data:
    outputs.append(deduce_numbers(signal, display))
print(sum(outputs))
# part 2: 989396


989396
